### PR TITLE
feat(ui): Disable user selection of data when user does not have export permissions

### PR DESCRIPTION
**Pull request by MialLewis**
### SUMMARY
Disables the user selection of data from the sqllab `resultTable`, `plugin-chart-pivot-table` and `plugin-chart-table` if the user does not have the `can_export_csv` / `can_csv` role. This is based off: https://github.com/apache/superset/pull/28429.

Relevant discussion: https://github.com/apache/superset/discussions/29078

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Required feature flags:
- [x] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

**Comment by MialLewis** [comment]
@rusackas I'm pinging you as you merged a related PR https://github.com/apache/superset/pull/28429.

I'm working to disable user select for data, which is required for our sensitive data. The more blockers we can put in the place of data export the better, even if they can be circumvented.

We are willing to fork `superset` to do this, but getting it merged would obviously be better! Do you think there would be any appetite for this change, willing to put behind a feature flag if necessary.

Also, any suggestions/examples of how best to use the permissions model inside the frontend plugins?

Thanks!

**Comment by MialLewis** [comment]
This is ready for someone to take a look. I'm very open to feedback as this is the first time I've used react. I do have very limited time to work on this though, it's already serving our purposes as it.

Thanks!

File: superset-frontend/src/components/FilterableTable/index.tsx [hunk 1]
```diff
@@ -249,11 +257,16 @@ const FilterableTable = ({
       }),
   }));
 
+  const canExportData = useSelector((state: RootState) =>
+    findPermission('can_export_csv', 'SQLLab', state.user?.roles),
+  );
+
```
Comment (justinpark) [inline superset-frontend/src/components/FilterableTable/index.tsx:263]: The FilterableTable is a common component, so accessing the Redux user state to check permissions is not an appropriate approach. (Additionally, if FilterableTable is used outside of SQLLab, this permission logic would be incorrect.)
Therefore, it would be more appropriate to add a prop to FilterableTable (e.g., disableTextSelection?: boolean) and include this logic in SqlLab/components/ResultSet, passing it to FilterableTable.

File: superset-frontend/src/components/Chart/Chart.jsx [hunk 2]
```diff
@@ -156,6 +158,11 @@ class Chart extends PureComponent {
     super(props);
     this.handleRenderContainerFailure =
       this.handleRenderContainerFailure.bind(this);
+    this.canExportData = findPermission(
+      'can_csv',
+      'Superset',
+      this.props.user?.roles,
+    );
```
Comment (justinpark) [inline superset-frontend/src/components/Chart/Chart.jsx:165]: For similar reasons [below](https://github.com/apache/superset/pull/29654/files#r1746286537), adding a user state to ChartContainer and ChartRenderer and including this kind of logic would not be flexible in terms of component reusability. Therefore, it would be more appropriate to handle this at the dashboard or explore component level. (Additionally, in the case of the findPermission logic, if there are many charts in the dashboard, repeating the same n+1 search logic for each chart in the user permission list would lead to inefficiency.)

In other words, this permission logic should be included in DashboardBuilder and passed as a property value to DashboardComponent, allowing access through `this.props.canExportData`.
Similarly, if it's included in ExploreChartPanel and passed as a property to ChartContainer, it can be accessed in the same way.

File: superset-frontend/packages/superset-ui-core/src/chart/models/ChartProps.ts [hunk 3]
```diff
@@ -105,6 +105,8 @@ export interface ChartPropsConfig {
   inputRef?: RefObject<any>;
   /** Theme object */
   theme: SupersetTheme;
+  /** Set chart data selectable by user */
+  dataSelectionMode?: string;
```
Comment (justinpark) [inline superset-frontend/packages/superset-ui-core/src/chart/models/ChartProps.ts:109]: It seems more appropriate to pass a more shareable value like canExportData rather than sending specific data selection values, and then set it according to each component's needs.

```suggestion
  canExportData: boolean;
```

File: superset-frontend/plugins/plugin-chart-pivot-table/src/plugin/transformProps.ts [hunk 4]
```diff
@@ -81,6 +81,7 @@ export default function transformProps(chartProps: ChartProps<QueryFormData>) {
     filterState,
     datasource: { verboseMap = {}, columnFormats = {}, currencyFormats = {} },
     emitCrossFilters,
+    dataSelectionMode,
```
Comment (justinpark) [inline superset-frontend/plugins/plugin-chart-pivot-table/src/plugin/transformProps.ts:84]:
```suggestion
    canExportData,
```

File: superset-frontend/plugins/plugin-chart-pivot-table/src/plugin/transformProps.ts [hunk 5]
```diff
@@ -176,5 +177,6 @@ export default function transformProps(chartProps: ChartProps<QueryFormData>) {
     onContextMenu,
     timeGrainSqla,
     allowRenderHtml,
+    dataSelectionMode,
```
Comment (justinpark) [inline superset-frontend/plugins/plugin-chart-pivot-table/src/plugin/transformProps.ts:180]: Here is the place to pass the appropriate data selection mode value for the given viz plugin.
```suggestion
    dataSelectionMode: canExportData ? 'auto' : 'none'
```

**Comment by rusackas** [comment]
Running CI. Ping us if it gets stuck!

File: superset-frontend/src/components/Chart/Chart.jsx (hunk 2 above)
Comment (MialLewis) [reply superset-frontend/src/components/Chart/Chart.jsx]: With regards to dashboards, there is already a property `supersetCanCSV` passed through to `dashboard/components/gridCompnents/Chart.jsx`, so we can pass that to `ChartContainer`: `canExportData={supersetCanCSV}`

**Comment by MialLewis** [comment]
Thanks for the review @justinpark, I've made the changes you suggested.

@rusackas would you mind running the CI again, it wasn't quite ready last time. I've run prettier locally so hopefully that will stop the failures.

**Comment by MialLewis** [comment]
@rusackas could you re-run the CI please?

I've built the js packages locally and see no test failures now. I haven't been able to run the cypress tests however as I see a `message": "400 Bad Request: The CSRF token is missing` when cypress attempts to login.

**Comment by rusackas** [comment]
Running CI 🤞

**Comment by rusackas** [comment]
Hi there... sorry this thing seems to have fallen under our radar. We're trying to thin the backlog of unmergeable things, so we can keep a better handle on the situation and have this happen _less_ often. I'm not sure how hard this is to rebase, but if you want to give it a go, we can try to get it merged (finally). I'll convert it to a draft in the meantime, and nowadays when you mark it ready for review, it'll get our attention again.

**Comment by rusackas** [comment]
This draft hasn't been updated in >90 days, so it will be closed. Please feel free to reopen it at any time, or ping us on this thread for further help.
